addons/pkg/util/condition_util.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+)
+
+// SummarizeAppConditions summarizes the provided conditions slice into a single condition with the following logic:
+// - If there be any 'Reconciling' condition type with status 'True', the summary condition type will be 'Reconciling'
+// - Otherwise, if there be any 'ReconcileFailed' condition type with status 'True', the summary condition type will be 'ReconcileFailed'
+// - Otherwise, if there be any 'ReconcileSucceeded' condition type with status 'True', the summary condition type will be 'ReconcileSucceeded'
+// - Otherwise, if there be any 'Deleting' condition type with status 'True', the summary condition type will be 'Deleting'
+// - Otherwise, if there be any 'DeleteFailed' condition type with status 'True', the summary condition type will be 'DeleteFailed'
+// - Otherwise, the summary condition type will be 'UnknownCondition'
+// Note that ReconcileFailed|Reconciling|ReconcileSucceeded|Deleting|DeleteFailed|DeleteSucceeded are mutually exclusive
+func SummarizeAppConditions(conditions []v1alpha1.AppCondition) v1alpha1.AppCondition {
+	summaryCondition := v1alpha1.AppCondition{}
+
+	wantedCondTypes := []v1alpha1.AppConditionType{
+		v1alpha1.Reconciling,
+		v1alpha1.ReconcileFailed,
+		v1alpha1.ReconcileSucceeded,
+		v1alpha1.Deleting,
+		v1alpha1.DeleteFailed,
```

Comment:
The list of conditions is slightly confusing for me because it seems that we are using condition type in mixed ways:
- we are using it to express both actions in progress and actions outcomes; those seem to me two states of the same condition vs being two different conditions
- we have a mixed positive/negative polarity that can be confusing (even if this exists in many K8s objects)

Reply:
Well these are Carevel kapp controller's PackageInstall condition types https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/apis/kappctrl/v1alpha1/status.go#L26. As the goal was to reflect pkgi condition type, I had to use these.